File: homy/integrations/immich/widgets/stats/stats.py
"""Immich library stats widget."""
import logging

from flask import jsonify, request

from homy.cache import cache_key, should_bypass_cache, widget_cache
from homy.integration_widget_util import get_integration_for_widget, get_widget_config
from homy import integration_widget_fetch as iwf

logger = logging.getLogger(__name__)

WIDGET = {
    'type': 'immich_stats',
    'name': 'Immich Library',
    'integration_types': ['immich'],
    'default_size_x': 6,
    'default_size_y': 4,
    'icon': 'image',
    'config_schema': {
        'integration_id': {'type': 'text', 'label': 'Integration ID', 'default': ''},
        'storage_warning_gb': {
            'type': 'text',
            'label': 'Speicher-Warnung ab (GB)',
            'default': '100',
        },
        'show_version': {
            'type': 'select',
            'label': 'Version anzeigen',
            'options': ['Ja', 'Nein'],
            'default': 'Nein',
        },
    },
}

CACHE_NS = 'immich_stats'
CACHE_TTL = 300


def register(app):
    @app.route('/api/immich_stats/data', methods=['GET'], endpoint='widget_immich_stats_data')
    def immich_stats_data():
        widget_id = request.args.get('widget_id')
        if not widget_id:
            return jsonify({'error': 'Bad Request', 'message': 'widget_id required'}), 400

        wcfg = get_widget_config(widget_id)
        if wcfg is None:
            return jsonify({'error': 'Not Found'}), 404

        integration, config, err, code = get_integration_for_widget(widget_id, ['immich'])
        if err:
            return jsonify({'configured': False, 'online': False, 'message': err}), code

        ck = cache_key(CACHE_NS, widget_id)
        if not should_bypass_cache():
            cached = widget_cache.get(ck)
            if cached is not None:
                return jsonify(cached)

        try:
            result = iwf.fetch_immich_stats(config)
            try:
                result['storage_warning_gb'] = float(wcfg.get('storage_warning_gb') or 100)
            except (TypeError, ValueError):
                result['storage_warning_gb'] = 100.0
            result['show_version'] = wcfg.get('show_version', 'Nein') == 'Ja'
            result['configured'] = True
            result['online'] = True
            widget_cache.set(ck, result, ttl=CACHE_TTL)
            return jsonify(result)
        except Exception as exc:
            logger.warning('immich_stats failed: %s', exc)
            return jsonify({'configured': True, 'online': False, 'message': str(exc)}), 200

File: homy/integrations/immich/widgets/stats/stats.js
window.WidgetRegistry.register('immich_stats', {
    async render(container, widgetData, config) {
        container.innerHTML = '<div class="widget-loading"><div class="spinner"></div></div>';
        try {
            const data = await API.request(`/api/immich_stats/data?widget_id=${widgetData.id}`);
            container.innerHTML = '';
            if (!data.online) {
                container.innerHTML = `<div class="widget-error text-center" style="padding:12px;">${data.message || 'Not available'}</div>`;
                return;
            }

            const usageGb = data.usage_gb ?? 0;
            const warnGb = data.storage_warning_gb ?? 0;
            const storPct = warnGb > 0 ? Math.min(100, Math.round((usageGb / warnGb) * 100)) : 0;
            const barCls = storPct >= 90 ? 'im-bar-danger' : storPct >= 70 ? 'im-bar-warn' : '';

            const versionHtml = data.show_version && data.version
                ? `<span class="im-version">v${data.version}</span>` : '';

            const storageHtml = `<div class="im-storage">
                <div class="im-storage-label">
                    <span>Speicher</span>
                    <span>${usageGb} GB${warnGb ? ` / ${warnGb} GB` : ''}</span>
                </div>
                <div class="im-storage-track"><div class="im-storage-fill ${barCls}" style="width:${warnGb ? storPct : 100}%"></div></div>
            </div>`;

            container.innerHTML = `<div class="im-container">
                <div class="im-header">
                    <i data-lucide="image" class="im-header-icon"></i>
                    <span class="im-title">Immich</span>
                    ${versionHtml}
                </div>
                <div class="im-counts">
                    <div class="im-count"><i data-lucide="camera" class="im-count-icon"></i><div><span class="im-count-val">${(data.photos || 0).toLocaleString()}</span><span class="im-count-label">Fotos</span></div></div>
                    <div class="im-count"><i data-lucide="film" class="im-count-icon"></i><div><span class="im-count-val">${(data.videos || 0).toLocaleString()}</span><span class="im-count-label">Videos</span></div></div>
                    <div class="im-count"><i data-lucide="hard-drive" class="im-count-icon"></i><div><span class="im-count-val">${usageGb}</span><span class="im-count-label">GB</span></div></div>
                </div>
                ${storageHtml}
            </div>`;
            window.refreshLucideIcons?.(container);
        } catch (err) {
            container.innerHTML = `<div class="widget-error"><i data-lucide="alert-triangle"></i><span>${err.message}</span></div>`;
            window.refreshLucideIcons?.(container);
        }
    },
});
